dask_kubernetes/objects.py

```diff
@@ -229,26 +229,56 @@ def clean_pod_template(pod_template):
     else:
         pod_template.spec.tolerations.extend(tolerations)
 
-    affinity = client.V1PreferredSchedulingTerm(
-        weight=100,
-        preference=client.V1NodeSelectorTerm(
+    # add default node affinity to k8s.dask.org/node-purpose=worker
+    if match_node_purpose is not "ignore":
+        # for readability
+        affinity = pod_template.spec.affinity
+
+        if affinity is None:
+            affinity = client.V1Affinity()
+        if affinity.node_affinity is None:
+            affinity.node_affinity = client.V1NodeAffinity()
+
+        # a common object for both a preferred and a required node affinity
+        node_selector_term = client.V1NodeSelectorTerm(
             match_expressions=[
                 client.V1NodeSelectorRequirement(
-                    key='k8s.dask.org/node-purpose',
-                    operator='In',
-                    values=['worker']
+                    key="k8s.dask.org/node-purpose", operator="In", values=["worker"]
                 )
             ]
         )
-    )
-    
-    if pod_template.spec.affinity is None:
-        pod_template.spec.affinity = client.V1Affinity(
-                node_affinity=client.V1PodAntiAffinity(
-                    preferred_during_scheduling_ignored_during_execution=[affinity]
+
+        if match_node_purpose == "require":
+            if (
+                affinity.node_affinity.required_during_scheduling_ignored_during_execution
+                is None
+            ):
+                affinity.node_affinity.required_during_scheduling_ignored_during_execution = client.V1NodeSelector(
+                    node_selector_terms=[]
                 )
+            affinity.node_affinity.required_during_scheduling_ignored_during_execution.node_selector_terms.append(
+                node_selector_term
             )
-    else:
-        print('TODO: need to handle the various partial states here')
+        elif match_node_purpose == "prefer":
+            if (
+                affinity.node_affinity.preferred_during_scheduling_ignored_during_execution
+                is None
+            ):
+                affinity.node_affinity.preferred_during_scheduling_ignored_during_execution = (
+                    []
+                )
```

Comment:
The assignment above should be an array I think. If we could say the type of elements it should be an array of V1PreferredSChedulingTerm objects, but as we cannot type the arrays element it should simply be an array.

I understand it currently as it is a tuple with one element: an array. I'm not confident on how the parenthesis are interpreted though.

Reply:
This evaluates to an empty list.

```ipython
In [1]: x = ([])

In [2]: x
Out[2]: []
```

As you say, we can't type this further. AFAIK, there isn't a way to instantiate an empty list of `V1PreferredSchedulingTerm` objects, like we do for the `V1NodeSelector`s above.

Reply:
Ah okay! Technically then all good!!

I just learned that `()` is a 0-tuple or a tuple with 0 elements, and ("test",) is a tuple with 1 element, and ("test", "more") is a tuple with two elements, but ("something") is just "something".